File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    // do a in order traversal to store all the nodes in an array
    // index the k - 1 element
    int kthVal = 0;
    int cnt = 0;
    public int kthSmallest(TreeNode root, int k) {
        helper(root, k);
        return kthVal;
    }

    public void helper(TreeNode root, int target) {
        if (root == null) {
        }
        else {
            helper(root.left, target);
            cnt++;
            System.out.println(root.val + " " + cnt);
            if (cnt == target) {kthVal = root.val;}
            helper(root.right, target);
        }
    }
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        if not root.left and not root.right:
            return 1

        countGood = 0

        def dfs(node, maxVal):
            nonlocal countGood
            
            if not node:
                return
            
            newMax = maxVal
            if node.val >= maxVal:
                countGood += 1
                newMax = node.val
            
            dfs(node.left, newMax)
            dfs(node.right, newMax)

        
        dfs(root, root.val)
        return countGood